ReferenceError doesn't include variable name in error message

colelao opened the pull request:
#### 412a6c047fa6e7ba2062e1bbd8afc707931b7cab
<pre>
ReferenceError doesn&apos;t include variable name in error message
<a href="https://bugs.webkit.org/show_bug.cgi?id=275143">https://bugs.webkit.org/show_bug.cgi?id=275143</a>
<a href="https://rdar.apple.com/129659300">rdar://129659300</a>

Reviewed by Keith Miller.

This improves the JSC `ReferenceError` caused by a tdz error by including the variable name in the message. Both Chrome and Firefox do this.
This is achieved by emitting `ExpressionInfo` in the bytecode that corresponds to the identifier. The `ExpressionInfo` is
then used when there is a tdz error. A majority of the changes are fixing tests to follow this format, and the most important changes are
in `NodesCodegen.cpp`

Canonical link: <a href="https://commits.webkit.org/300119@main">https://commits.webkit.org/300119@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a74c016dde76a90f9b6eff90734341e27f3cba3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121431 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31787 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127879 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/73514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49707 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/92238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/73514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124383 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/33400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/108789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/72914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/32414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26944 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71452 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/113561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/102894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130706 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/119951 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48359 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36769 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/100828 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48727 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105003 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100735 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25528 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/46145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/24215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/48217 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53930 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150113 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/47689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38194 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51035 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/49371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
This looks concerning since now all check_tdz bytecode gets larger only for the exceptional case. I think we should have a mechanism to extract this kind of additional error information only when throwing an error instead of increasing the bytecode size (IIRC V8 does that).

MenloDorian commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -1244,6 +1244,7 @@
 op :check_tdz,
     args: {
         targetVirtualRegister: VirtualRegister,
+        identifier?: VirtualRegister,
```
This is going to increase bytecode size, and the only gain appears to be just for an error message.  Have you measured if this increase is material or not on benchmarks?

kmiller68 commented on Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
```diff
@@ -340,6 +340,11 @@ JSObject* createErrorForInvalidGlobalVarDeclaration(JSGlobalObject* globalObject
     return createTypeError(globalObject, makeString("Can't declare global variable '"_s, ident.string(), "': global object must be extensible"_s));
 }
 
+JSObject* createTDZError(JSGlobalObject* globalObject, const Identifier& ident)
```
Doesn't seem like this needs to be an `Identifier` it probably already is in most cases but we should make this just a `StringView`. It's pretty expensive to make a new AtomString so we should avoid it if it's unnecessary.

kmiller68 reviewed (commented):
Overall, looks good. One minor change though.

kmiller68 commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -306,8 +307,16 @@ JSC_DEFINE_COMMON_SLOW_PATH(slow_path_check_tdz)
+        int expressionStop = info.divot + info.endOffset;
+
+        RefPtr provider = codeBlock->source().provider();
+        auto ident = provider->getRange(expressionStart, expressionStop).toString();
```
I don't think you need to do `.toString()` here. Although you might need to annotate `LIFETIME_BOUND` on `getRange`

kmiller68 commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -153,7 +153,8 @@ bool JSModuleNamespaceObject::getOwnPropertySlotCommon(JSGlobalObject* globalObj
         if (!value) {
-            throwVMError(globalObject, scope, createTDZError(globalObject));
+            RefPtr uid = propertyName.uid();
+            throwVMError(globalObject, scope, createTDZError(globalObject, JSC::Identifier::fromUid(vm, uid.get())));
```
I think this can just be `uid.get()`? Is there a reason to wrap it into an Identifier anymore? You should also be in the `JSC::` namespace so you don't need that.

kmiller68 reviewed (approved):
nice, r=me, thanks!

colelao committed:
ReferenceError doesn't include variable name in error message
https://bugs.webkit.org/show_bug.cgi?id=275143
rdar://129659300

Reviewed by Keith Miller.

This improves the JSC `ReferenceError` caused by a tdz error by including the variable name in the message. Both Chrome and Firefox do this.
This is achieved by emitting `ExpressionInfo` in the bytecode that corresponds to the identifier. The `ExpressionInfo` is
then used when there is a tdz error. A majority of the changes are fixing tests to follow this format, and the most important changes are
in `NodesCodegen.cpp`

Canonical link: https://commits.webkit.org/300119@main
